reflex/app.py

```diff
@@ -94,8 +94,20 @@ def __init__(self, *args, **kwargs):
         Args:
             *args: Args to initialize the app with.
             **kwargs: Kwargs to initialize the app with.
+
+        Raises:
+            ValueError: If there are multiple client subclasses of rx.State(Subclasses of rx.State should consist
+                        of the DefaultState and the client app state).
         """
         super().__init__(*args, **kwargs)
+        state_subclasses = State.__subclasses__()
+
+        # only the default state and the client state should be allowed as subclasses
+        if len(state_subclasses) > 2:
+            raise ValueError(
+                "rx.State has been subclassed multiple times. Only one subclass is allowed"
+            )
+        self.state = state_subclasses[-1]
```

Comment:
> since users aren't allowed to define multiple subclasses of `rx.State`

i thought this was only the case for implicit determination of the `State` subclass. If the user explicitly provides a `state=` kwarg, then does it really matter if they've defined multiple subclasses of `rx.State`? I suppose it might not do what they want, but I don't necessarily want to limit advanced use cases, like implementations that have multiple `rx.State` subclasses, but then only pick one at runtime, even though they are all technically loaded and available (and would fail the `len(state_subclasses) < 2` check).

Reply:
Ah right, the current design only supports having one subclass(child excluding the default state) of `rx.State` and n grandchildren or descendants. Supporting n children of `rx.State` will require fundamental modifications. I'm still curious, are there some advanced cases where having multiple children of `rx.State` will be the best case? @masenf @picklelo

Reply:
i'm not clear on what is broken if there are multiple subclasses of `rx.State`?

the advanced use cases i'm thinking of are around testing mostly, where you'll have multiple test cases, maybe each defining their own `rx.State` subclasses and passing those off as `App(state=Test1State)`. It would be nice to be able to support that without _having_ to patch the `__subclasses__` return.